core/src/main/java/org/apache/iceberg/io/ContentCache.java

```diff
@@ -0,0 +1,310 @@
+      } else {
+        return length;
+      }
+    }
+
+    /**
+     * Opens a new {@link SeekableInputStream} for the underlying data file, either through cache or
+     * through the inner FileIO.
+     *
+     * <p>If data file is not cached yet, and it can fit in the cache, the file content will be
+     * cached first before returning a {@link ByteBufferInputStream}. Otherwise, return a new
+     * SeekableInputStream from the inner FIleIO.
+     *
+     * @return a {@link ByteBufferInputStream} if file exist in the cache or can fit in the cache.
+     *     Otherwise, return a new SeekableInputStream from the inner FIleIO.
+     */
+    @Override
+    public SeekableInputStream newStream() {
+      try {
+        // read-through cache if file length is less than or equal to maximum length allowed to
+        // cache.
+        if (getLength() <= contentCache.maxContentLength()) {
+          return cachedStream();
+        }
+
+        // fallback to non-caching input stream.
+        return wrappedInputFile().newStream();
+      } catch (FileNotFoundException e) {
+        throw new NotFoundException(
+            e, "Failed to open input stream for file %s: %s", location, e.toString());
+      } catch (IOException e) {
+        throw new RuntimeIOException(
+            e, "Failed to open input stream for file %s: %s", location, e.toString());
+      }
+    }
+
+    @Override
+    public String location() {
+      return location;
+    }
+
+    @Override
+    public boolean exists() {
+      CacheEntry buf = contentCache.getIfPresent(location);
+      return buf != null || wrappedInputFile().exists();
+    }
+
+    private CacheEntry cacheEntry() {
+      long start = System.currentTimeMillis();
+      try (SeekableInputStream stream = wrappedInputFile().newStream()) {
+        long fileLength = getLength();
+        long totalBytesToRead = fileLength;
+        List<ByteBuffer> buffers = Lists.newArrayList();
+
+        while (totalBytesToRead > 0) {
+          // read the stream in 4MB chunk
+          int bytesToRead = (int) Math.min(BUFFER_CHUNK_SIZE, totalBytesToRead);
+          byte[] buf = new byte[bytesToRead];
+          int bytesRead = IOUtil.readRemaining(stream, buf, 0, bytesToRead);
+          totalBytesToRead -= bytesRead;
+
+          if (bytesRead < bytesToRead) {
+            // Read less than it should be, possibly hitting EOF. Abandon caching by throwing
+            // IOException and let the caller fallback to non-caching input file.
+            throw new IOException(
+                "Expected to read "
+                    + fileLength
+                    + " bytes, but only "
+                    + (fileLength - totalBytesToRead)
+                    + " bytes "
+                    + "read.");
+          } else {
+            buffers.add(ByteBuffer.wrap(buf));
+          }
+        }
+
+        CacheEntry newEntry = new CacheEntry(fileLength - totalBytesToRead, buffers);
```

Comment:
`totalBytesToRead` should always be zero so this should just be `fileLength`

Reply:
Done.